src/runtime/gc_list.go

```diff
@@ -0,0 +1,460 @@
+// +build gc.list
+
+// This memory manager is partially inspired by the memory allocator included in avr-libc.
+// Unlike avr-libc malloc, this memory manager stores an allocation list instead of a free list.
+// There is an overhead of 2 pointer-widths/allocation (up from avr-libc's 1 pointer-width/allocation).
+// The allocation list is stored in ascending address order.
+// The set of free spans is implicitly derived from the gaps between allocations.
+// This allocator has an effective allocation complexity of O(n) and worst-case GC complexity of O(n^2).
+// Due to architectural quirks, as well as tiny heaps, this should almost always be faster than the standard conservative collector on AVR.
+
+package runtime
+
+import (
+	"internal/task"
+	"runtime/interrupt"
+	"unsafe"
+)
+
+// Set gcDebug to true to print debug information.
+const (
+	gcDebug   = false   // print debug info
+	gcAsserts = gcDebug // perform sanity checks
+)
+
+// setHeapEnd is called to expand the heap. The heap can only grow, not shrink.
+func setHeapEnd(newHeapEnd uintptr) {
+	if gcAsserts && newHeapEnd <= heapEnd {
+		panic("gc: setHeapEnd didn't grow the heap")
+	}
+
+	heapEnd = newHeapEnd
+
+	// There is no other metadata to update since the free space is just computed from the heap bounds and the allocations list.
+}
+
+// Initialize the memory allocator.
+func initHeap() {
+	// No initialization is required: the only state is the allocation list, which starts empty.
+	// This function must exist anyway since other GCs need initialization.
+	// It should be inlined and optimized away by the compiler.
+}
+
+var gcRunning bool
+
+// alloc tries to find some free space on the heap, possibly doing a garbage
+// collection cycle if needed. If no space is free, it panics.
+//go:noinline
+func alloc(size uintptr) unsafe.Pointer {
+	if gcAsserts && gcRunning {
+		runtimePanic("alloc while GC is running")
+	}
+
+	var ranGC bool
+tryAlloc:
+	// Search for available memory.
+	alloc := findMem(size)
+	if alloc == nil {
+		// There is no free span large enough for the allocation.
+
+		if ranGC {
+			// Even after running the GC, there is not enough memory.
+			if growHeap() {
+				// The heap was able to grow to fit more data.
+				// Try allocating again.
+				// When doing a huge allocation this will loop until the heap is big enough.
+				goto tryAlloc
+			}
+
+			// There is no more available memory.
+			runtimePanic("out of memory")
+		}
+
+		// Run the garbage collector and try again.
+		GC()
+		ranGC = true
+		goto tryAlloc
+	}
+
+	// Zero the allocation.
+	ptr := unsafe.Pointer(uintptr(unsafe.Pointer(alloc)) + unsafe.Sizeof(allocHeader{}))
+	memzero(ptr, size)
+
+	return ptr
+}
+
+// findMem searches the heap for a free span large enough to contain an allocation of the specified size.
+// If there are no sufficiently large free spans available, this returns nil.
+func findMem(size uintptr) *allocHeader {
```

Comment:
Two things I realize now:

- The memory should be aligned. You can use the `align` function for this. This is a no-op on AVR and should be optimized away entirely. On other platforms, it is necessary for correct functioning.
- You might want to use a `zeroSizedAlloc` global (like in gc_conservative.go) to return for when the size is zero. This does sometimes happen in practice, although it is rare. The Go spec explicitly allows this:
  > Pointers to distinct zero-size variables may or may not be equal.
  
  On the other hand, it's rare and costs RAM/flash, so it would be reasonable to leave this out.

Reply:
Adding the 0-sized alloc in because it only costs 1 byte of RAM (and saves 28 bytes of flash for some reason).

Reply:
I am not entirely sure what you mean with the alignment. Assuming that `heapStart` is aligned (which I picked up from the conservative collector), `findMem` should always return appropriately aligned data. The header preserves pointer/word alignment, and every subsequent `start` is corrected with an align call.